ui: add API for command filtering for embedders

cdamus opened the pull request:
In some applications that embed the Perfetto UI, there are a number of commands that are not needed. For example, in an application that manages the opening of traces into instances of the Perfetto UI, there is no need for the commands that open and close traces. And other commands may overlap or conflict with other capabilities provided by the host application, which therefore may wish to exclude those commands.

Add an API to the Registry class to permit an embedding application to install a filter that quietly prevents registration of unwanted services. Employ this capability in the CommandManager.

For android-graphics/sokatoa#3871

cdamus committed:
ui: add API for command filtering for embedders

In some applications that embed the Perfetto UI, there are a number of
commands that are not needed. For example, in an application that
manages the opening of traces into instances of the Perfetto UI, there
is no need for the commands that open and close traces. And other
commands may overlap or conflict with other capabilities provided by the
host application, which therefore may wish to exclude those commands.

Add an API to the Registry class to permit an embedding application to
install a filter that quietly prevents registration of unwanted
services. Employ this capability in the CommandManager.

Signed-off-by: Christian W. Damus <cdamus.ext@eclipsesource.com>

colin-grant-work commented on ui/src/base/registry.ts:
```diff
@@ -85,8 +101,14 @@ export class Registry<T> {
+        this.keyFilter = parent.keyFilter;
+      }
+
+      override set filter(filter: ((key: string) => boolean) | undefined) {
+        this.parent.filter = filter;
+        super.filter = filter;
       }
```
For our use case, this makes sense, but I'm not sure that propagating the setting up to the parent (without the parent then propagating it back down to all children?) would make sense as a general API. In practice, do we need this behavior, i.e. do we end up interacting with a child such that we need to make sure that when we set the filter on that child, it makes its way back up to the root Registry?

cdamus replied:
We do need this behaviour. Most of the commands that are registered on the parent (`AppImpl`-level, not open-Trace-level) are inapplicable and/or dangerous in Sokatoa.

We could require an application to separately set a filter on the parent if you prefer that. We certainly have access to it.

colin-grant-work replied:
If we have direct access to the registries where we need set filters, then I think removing child -> parent propagation would assuage any concerns I have. I think it's the inconsistency between child construction time (propagation parent -> child) and post-construction time (propagation child -> parent) that makes me a little queasy.

cdamus replied:
Agreed. It'll be cleaner this way.

colin-grant-work commented on ui/src/base/registry.ts:
```diff
@@ -36,8 +37,23 @@ export class Registry<T> {
+    [...this.registry.keys()]
+      .filter((key) => !this.keyFilter(key))
+      .forEach((key) => this.registry.delete(key));
+  }
```
This also seems a bit odd as API in the following scenario:
 1. Registrants register a variety of items.
 2. Someone sets a filter that allows only subset A through.
 3. Someone sets a filter that allows only subset B through.
 4. Now we end up with only the intersection of subsets A and B, but the filterer at step (3) might reasonably want to get everything in subset B that was ever registered, since the filter for subset A was never what they were interested in (or knew about, or could control, for whatever reason).

That makes me wonder whether it might make sense (and be possible, of course?) to limit what the registry exposes while retaining a reference to the original registrations? Something like 

```
/** retains everything */
registry
/** target for retrieval: contains the subset of items matching current filter */
filteredRegistry
```

cdamus replied:
The purpose is that the "someone" should be an application that embeds Perfetto and is the only authority for such filters. And so wouldn't be inconsistent about it. Nothing in this PR made that clear, of course.

Perhaps this should be a write-once property. Attempts to change the filter after it has been set should throw an error. And then it should be a `setFilter()` method and not a `set filter` prop.

colin-grant-work replied:
Set once behavior would make the expectations explicit 👍

colin-grant-work commented on ui/src/base/registry_unittest.ts:
```diff
@@ -176,4 +207,60 @@ describe('Hierarchical (child) registries', () => {
+      expect(() => child.get('beta')).toThrow();
+    });
+
+    test('setting filter on child cascades to parent and prunes both registries', () => {
+      const parent = Registry.kindRegistry<Registrant>();
+      const a: Registrant = {kind: 'a', n: 1};
+      const b: Registrant = {kind: 'b', n: 2};
+      parent.register(a);
+      parent.register(b);
+
+      const child = parent.createChild();
```
The scenario of interest here would be something like

```typescript
const parent = Registry.kindRegistry<Registrant>();
const childA = parent.createChild();
const childB = parent.createChild();

childA.filter = (candidate) => isCoolEnough(candidate);
// Should childB be affected, given that the parent is?
```

cdamus replied:
It depends on what you mean by "affected."

#### Assuming that filtering is not inherited, that filters are separately installed at every level of registry delegation, then: 

A child should only inherit from a parent the registrations that the parent itself publishes to the world. But the child can override those registrations.

So, if the parent filters out an uncool registration, then the child should not see it. But if the child does not have the cool filter, then the child should be able to add its own registration of an uncool entry.

#### Assuming that filtering is inherited, then

As always, a child cannot delegate the look-up of an uncool entry to the parent if the parent has a cool filter.

*Moreover*, a child should not be able to register an uncool entry if the parent has a cool filter.

cdamus committed:
Address review comments

- don't push child filter to parent
- dynamically inherit parent filter to child
  in addition to child filter
- make the filter write-once and be explicit about its purpose for
  embedder use cases

Signed-off-by: Christian W. Damus <cdamus.ext@eclipsesource.com>

cdamus commented:
Thanks for the review @colin-grant-work . Commit 848a0a301e54de74a5102aa3f3a5d4df49fedc66 should address your comments.

colin-grant-work reviewed (approved):
Thanks for the changes. LGTM.
